multi_imbalance/resampling/SOUP.py

```diff
@@ -0,0 +1,104 @@
+from collections import Counter, defaultdict
+from sklearn.neighbors import NearestNeighbors
+from sklearn.utils import shuffle
+
+import numpy as np
+
+
+class SOUP(object):
+    """
+    Similarity Oversampling and Undersampling Preprocessing (SOUP) is an algorithm that equalizes number of samples
+    in each class. It also takes care of the similarity between classes, which means that it removes samples from
+    majority class, that are close to samples from the other class and duplicate samples from th minority classes,
+    which are in the safest area in space
+    """
+
+    def __init__(self, k=9):
+        self.k = k
+        self.neigh_clf = NearestNeighbors(n_neighbors=self.k)
+        self.quantities, self.goal_quantity = [None] * 2
+
+    def fit_transform(self, X, y):
+        """
+
+        Parameters
+        ----------
+        X two dimensional numpy array (number of samples x number of features) with float numbers
+        y one dimensional numpy array with labels for rows in X
+
+        Returns
+        -------
+        Resampled X (mean class quantity * number of unique classes), y (number of rows in X) as numpy array
+        """
+        result_X, result_y = list(), list()
+        self.neigh_clf.fit(X)
+        self.quantities = Counter(y)
+        self.goal_quantity = np.mean(list(self.quantities.values()), dtype=int)
+
+        for class_name, class_quantity in self.quantities.items():
+
+            class_safe_levels = self._construct_class_safe_levels(X, y, class_name)
+
+            if class_quantity <= self.goal_quantity:
+                temp_X, temp_y = self._oversample(X, y, class_safe_levels)
+            else:
+                temp_X, temp_y = self._undersample(X, y, class_safe_levels)
+
+            result_X.extend(temp_X)
+            result_y.extend(temp_y)
+
+        result_X, result_y = shuffle(result_X, result_y)
+        return np.array(result_X), np.array(result_y)
+
+    def _construct_class_safe_levels(self, X, y, class_name):
+        indices_in_class = [i for i, value in enumerate(y) if value == class_name]
+
+        class_safe_levels = defaultdict(float)
+        for sample_id in indices_in_class:
+            neighbours_indices = self.neigh_clf.kneighbors([list(X[sample_id])], return_distance=False)
+            neighbours_classes = y[neighbours_indices[0]]
+            neighbours_quantities = Counter(neighbours_classes)
```

Comment:
The body of the loop could possibly be extracted as a method

Reply:
I agree. I extracted but also decided to don't create unit tests for function only with sklearn knn and built-in Counter - nothing to test ;)